### PR TITLE
Load filesystem cache metadata asynchronously

### PR DESCRIPTION
**Pull request by danipozo**
ClickHouse can take anywhere from 10 to 30 minutes to restart when the filesystem cache has been populated with millions of files. This change aims to make the loading of the cache asynchronous to the boot process to avoid this operational hurdle.

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Load filesystem cache metadata asynchronously to boot process, in order to make restarts faster.

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
Please add a setting which will be disabled by default.

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 8de18ad6df1742849d48efcb6142a70342fdf997 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65736/8de18ad6df1742849d48efcb6142a70342fdf997/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Cloud fork sync (only for ClickHouse Inc. employees)</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:1148]: We should be able to deactivate this thread in `FileCache::deactivateBackgroundOperations`.

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:1418]: Ok, but also we need to do `join` on the loading thread. So better to remove `load_metadata_main_thread.detach();` and save the thread inside the class.

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:1418]:
```
if (load_metadata_main_thread.joinable())
    load_metadata_main_thread.join();
```

Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp]: Thanks

File: src/Interpreters/Cache/FileCache.cpp
Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp]: Changed to have `isInitialized()` return `is_fully_loaded`, while `assertInitialized()` still checks `is_initialized`

Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp]: Actually, probably we could do with only one variable, since the methods that call `assertInitialized()` should probably not be called until the cache is loaded

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:206]: We should not do `metadata.startup()` and `keep_up_free_space_ratio_task->schedule();` until metadata is loaded, otherwise data race or segfault is possible, because currently `loadMetadata` relies on the fact that cache is not accessed unless metadata is loaded, but the methods I mentioned could do cache access in the background threads which they launch. So probably we need to do all of initialize() inside a background thread, but FileStatus file should be created in advance in the main thread to prevent cache breakage.

Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp:206]: Thanks for the detailed comment, I'll look into this next week.
> currently loadMetadata relies on the fact that cache is not accessed unless metadata is loaded

Would changing this be enough to be able to use the cache even if partially initialized?

Comment (kssenii) [reply src/Interpreters/Cache/FileCache.cpp:206]: > Would changing this be enough to be able to use the cache even if partially initialized?

No, currently code of loadMetadata relies on the fact that noone used cache until it is fully initialized. There are certain places where we do not take required mutexes because we know that it is initialization code. For example this one https://github.com/ClickHouse/ClickHouse/blob/3ef05fd11afb7e5827fda98cb833486649b340e4/src/Interpreters/Cache/FileCache.cpp#L1354
and also we cannot add new cache entries while cache is in the progress of initialization as the code of `loadMetadata` is not protected agains this

File: src/Disks/IO/ReadBufferFromRemoteFSGather.cpp
Comment (danipozo) [reply src/Disks/IO/ReadBufferFromRemoteFSGather.cpp:82]: This would make queries (that read from object storage) fail if cache fails to initialize, right? I wonder if it would be better to emit a warning instead, as queries can be made reading from object storage directly, or at least have a setting to not throw in that case

Comment (kssenii) [reply src/Disks/IO/ReadBufferFromRemoteFSGather.cpp:82]: > This would make queries (that read from object storage) fail if cache fails to initialize, right?

Yes, this behaviour will be consistent with what it currently is.

>  wonder if it would be better to emit a warning instead, as queries can be made reading from object storage directly, or at least have a setting to not throw in that case

Yes, let's add a setting, but by default let's leave the behaviour as it is currently is - e.g. throw.

File: src/Interpreters/Cache/FileCache.cpp
Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp:206]: > So probably we need to do all of initialize() inside a background thread, but FileStatus file should be created in advance in the main thread to prevent cache breakage.

Implemented these changes in https://github.com/ClickHouse/ClickHouse/pull/65736/commits/544a40c72a5cdaf63c3e820c83f68e36bd13ab24

**Review by nikitamikhaylov — CHANGES_REQUESTED** [review]
^

File: src/Disks/IO/ReadBufferFromRemoteFSGather.cpp
Comment (danipozo) [reply src/Disks/IO/ReadBufferFromRemoteFSGather.cpp:82]: Added a setting for this

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:153]: No need to log it here, it should be logged once - when the init itself failed, otherwise it will pollute the logs.

Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp]: Removed this log as it is already logged in `initializeImpl()`

**Comment by danipozo** [comment]
As far as I can tell, the remaining CI failures are unrelated to the changes in this PR

**Comment by danipozo** [comment]
@kssenii is there anything else that I should check/fix before merging? I've looked at the latest CI reports and to the best of my understanding none of the failures are related to this PR

**Comment by kssenii** [comment]
@danipozo sorry, but I will check next week when I return from vacation.

**Comment by danipozo** [comment]
> @danipozo sorry, but I will check next week when I return from vacation.

Of course, enjoy!

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Interpreters/InterpreterSystemQuery.cpp
Comment (danipozo) [reply src/Interpreters/InterpreterSystemQuery.cpp]: Done

File: src/Interpreters/Cache/FileCache.cpp
Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp:147]: I've added a doc comment to the function saying that it can throw

File: src/Interpreters/Cache/FileCache.cpp
Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp]: Done

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:192]: All initialize() calls should wait if the first initialize() is still in progress, otherwise they will try to use cache and will get "Cache is not initialized" error

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:197]: Please fix identation, it should be 4 spaces, not 2

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:227]: We should not call loadMetadata() if base path did not exist, when we called initialize(). Because we will start background threads in vain. So we need to not call initializeImpl() from initialize() in lines 208-216 if we saw earlier that base path did not exist

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:207]: I'd rather leave is_initalized setting to true here, as it was before, now you moved to loadMetadataImpl(), which could lead to unnoticed error from this part of the code, while it is very important

File: src/Interpreters/Cache/FileCache.cpp
Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp]: Done this

File: src/Interpreters/Cache/FileCache.cpp
Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp]: Changed the location of the assignment

File: src/Interpreters/Cache/FileCache.cpp
Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp]: Done

File: src/Interpreters/Cache/FileCache.cpp
Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp]: I've substituted the guard using the atomic for a call to `callOnce`, so that all threads calling `initialize()` will wait until the first `initialize()` is finished if `load_metadata_asynchronously` is disabled.

In the case where it is enabled, I think all uses of the cache are guarded by `isInitialized()` calls, do you have some execution path in mind that could cause this error? I've also tried creating a lot of tables with the same cache and inserting in them concurrently with a sleep in `loadMetadata` and haven't been able to get that error.

Comment (kssenii) [reply src/Interpreters/Cache/FileCache.cpp]: We call initialize() for cache when we start disks on server startup, all disks are started one by one synchronously, so currently intialize() should not be called in parallel unless some other thread during startup accesses the cache in the meantime (cannot remember such case now), but anyway we should be prepared for this to happen. std::call_once is ok.

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:198]:
```suggestion
            if (!need_to_load_metadata)
                fs::create_directories(getBasePath());
```

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:214]: Exceptions from `initializeImpl` should also be caught and set to `init_exception`, both here and if load_metadata_asynchronously.

File: src/Interpreters/InterpreterSystemQuery.cpp
Comment (kssenii) [inline src/Interpreters/InterpreterSystemQuery.cpp:413]:
```suggestion
                        tryLogCurrentException(log, std::format("Cache {} is broken during initialization: ", cache_data->getSettings().toString()));
                        continue;
```

File: src/Interpreters/Cache/FileCache.cpp
Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp:214]: Exceptions from the call to `loadMetadata()` in `initializeImpl()` are handled that way inside `initializeImpl()`. The call to `metadata.startup()` is not handled that way as it was the previous behavior, but I guess it makes sense to include it in the `try` block too, I've added it

File: src/Interpreters/InterpreterSystemQuery.cpp
Comment (danipozo) [reply src/Interpreters/InterpreterSystemQuery.cpp:413]: Applied this in latest commit

File: src/Interpreters/Cache/FileCache.cpp
Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp]: Applied this (and set indentation to 4 spaces in my editor, sorry)

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [reply src/Interpreters/Cache/FileCache.cpp:214]: Indeed. But now there can be a data race on `init_exception`. Please add a mutex.

Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp:214]: What's the behavior that you'd expect in `isInitialized()`? I'm thinking it's good to avoid a data race on `init_exception`, but it should not block if  `init_mutex` is taken. WDYT about this?:
```cpp
        std::unique_lock lock(init_mutex, std::try_to_lock);
        if (lock)
        {
            /// Exception during initialization,
            /// make user aware that the cache is broken.
            if (init_exception && throw_on_cache_initialization_error)
            {
                std::rethrow_exception(init_exception);
            }
        }
        return is_initialized;
```

Comment (kssenii) [reply src/Interpreters/Cache/FileCache.cpp:214]: something like this
```
        if (is_initialized)
            return true;
        if (!throw_on_cache_initialization_error)
            return false;
        std::lock_guard(init_mutex);
        if (init_exception)
            std::rethrow_exception(init_exception);
        return is_initiailized;
```

Comment (kssenii) [reply src/Interpreters/Cache/FileCache.cpp:214]: `assertInitialized` anyway takes the lock as well: https://github.com/ClickHouse/ClickHouse/blob/7e9e03b9c67d14878fb1bbf8b8eb5aa047a2327a/src/Interpreters/Cache/FileCache.cpp#L157-L163

Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp:206]: I think we're overcomplicating `isInitialized()` with logic that is hardly expected in a getter. In the latest commit I've made a different proposal:
- `isInitialized()` simply returns `is_initialized`,
- added a method `initException()` that is a getter for `init_exception`.

Rationale for this proposal is:
- `isInitialized()` is simplified
- `throw_on_cache_initialization_error` setting is not necessary
- Code that only cares about whether the cache is initialized (`ReadBufferFromRemoteFSGather`) can use this method without handling exceptions
- Code that wants to handle the exception, such as `InterpreterSystemQuery`, can get the exception to do so

The only thing that it doesn't have is the behavior that you requested in https://github.com/ClickHouse/ClickHouse/pull/65736#discussion_r1657314071, could we reconsider this? I find it inconsistent that if queries work by reading from cloud storage directly while the cache is in process of initialization, they start failing once an error is found. And we are already logging an error in `initializeImpl()`.

If you still think it is necessary for queries to throw if the cache has failed to initialize, we can add code similar to the one in `InterpreterSystemQuery` to check and throw `init_exception`, possibly in a helper method like the one you proposed (that doesn't block on `init_mutex`, as that would render the `load_metadata_asynchronously` setting useless, queries would have to wait for the cache in any case)

Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp:200]: I've removed `initException()` since it could cause a data race if two threads want to throw the exception simultaneously

Comment (kssenii) [reply src/Interpreters/Cache/FileCache.cpp:200]: > The only thing that it doesn't have is the behavior that you requested in https://github.com/ClickHouse/ClickHouse/pull/65736#discussion_r1657314071, could we reconsider this? I find it inconsistent that if queries work by reading from cloud storage directly while the cache is in process of initialization, they start failing once an error is found. And we are already logging an error in initializeImpl().

If asynchronous loading if metadata is turned on - yes, it is ok not to rethrow the exception ever, so the user will always read from object storage. But in case it is disabled, which is the default, then we want to have old behaviour, that if cache failed to initialize on server startup, then it is visible to the user by default.

File: src/Core/SettingsChangesHistory.cpp
Comment (kssenii) [inline src/Core/SettingsChangesHistory.cpp:79]: Accidental change I guess

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [reply src/Interpreters/Cache/FileCache.cpp:200]: > If asynchronous loading if metadata is turned on - yes, it is ok not to rethrow the exception ever, so the user will always read from object storage.

In this case no need to worry about data race on init_exception, no mutex needed.

File: src/Interpreters/Cache/FileCache.cpp
Comment (kitaisreal) [inline src/Interpreters/Cache/FileCache.cpp:151]: @kssenii I tried to read previous comments but probably did not understand them.
Why do we want to change behaviour here ? 
The `FileCache` interface was previously designed naturally and can be used for async load and sync load.
Client before trying to use cache check if cache initialized using `isInitialized` method and if it returns false, fallback to reading method without cache. If client accidentally did not check `isInitialized` and tried to use cache, any `FileCache` method throws exception using `assertInitialized`. 
Throwing exception in `isInitialized` method does not look very obvious, because no one can expect that. What is the motivation ? 
Maybe for client it is better to manually use `assertInitialized` if he wants to get Exception.

File: src/Interpreters/Cache/FileCacheSettings.h
Comment (kitaisreal) [inline src/Interpreters/Cache/FileCacheSettings.h:35]: Before merge setting need to be enabled by default and all tests need to pass.
Then after review is approved setting needs to be disabled by default. And separate pull request created where setting is enabled by default.
So we can enable it by default in next version.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Interpreters/Cache/FileCache.cpp
Comment (kssenii) [reply src/Interpreters/Cache/FileCache.cpp:151]: Previously, before this PR, we would try to read from cache without checking `isInitialized` and will directly go to `assertInitialized`, which was the expectation that cache must always be initialized on server startup. But now in this PR we would not try to use cache for read-through/write-through cache if `isInitialized` returns `false` (because of the new check in ReadBufferFromRemoteFSGather). But it will return `false` in case initialization failed as well. So the previous behaviour changes - that with non-async cache initialization we will no longer catch that cache failed to be initialized. I want to preserve previous default behaviour, which is possible without throwing exception from `isInitialized` though, I was not forcing exactly this, but just preserving old behaviour.

File: src/Interpreters/Cache/FileCacheSettings.h
Comment (kssenii) [reply src/Interpreters/Cache/FileCacheSettings.h:35]: > Before merge setting need to be enabled by default and all tests need to pass.

we can also make it randomized similar to cache policies https://github.com/ClickHouse/ClickHouse/blob/18f1ec57cb33b439bc06bad8c580f6b25631153a/docker/test/stress/run.sh#L75-L90

File: src/Interpreters/Cache/FileCache.cpp
Comment (danipozo) [reply src/Interpreters/Cache/FileCache.cpp]: I've moved the logic to throw to a separate method `throwInitExceptionIfNeeded`

File: src/Interpreters/Cache/FileCacheSettings.h
Comment (kitaisreal) [reply src/Interpreters/Cache/FileCacheSettings.h]: Agree. Let's do it.

Comment (danipozo) [reply src/Interpreters/Cache/FileCacheSettings.h]: I've set this to enabled by default to check and fix CI. Checking crashing gtest at the moment. Once the CI is passing and before merging, I'll add a randomization for this setting, in order to catch as many failures as possible with this setting enabled now.

File: src/Interpreters/Cache/FileCache.cpp
Comment (kitaisreal) [inline src/Interpreters/Cache/FileCache.cpp:148]: First it is better to check `load_metadata_asynchronously` and return early.
Then use `std::lock_guard lock(init_mutex);` to avoid data races when other thread set `init_exception`.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

**Comment by danipozo** [comment]
CI failures now unrelated to the changes, as far as I can tell. This is with the setting enabled by default

File: src/Interpreters/TemporaryDataOnDisk.cpp
Comment (kssenii) [inline src/Interpreters/TemporaryDataOnDisk.cpp:68]: This might be a breaking change for those who rely only on cache for temporary data and do not have volume for tmp data defined in configuration... Let's disable async cache loading by default for now and add randomization of this setting for tests.

Comment (danipozo) [reply src/Interpreters/TemporaryDataOnDisk.cpp:68]: > Let's disable async cache loading by default for now

Yes, I just want to make sure the CI passes with it enabled by default, as per this comment https://github.com/ClickHouse/ClickHouse/pull/65736#discussion_r1699837710. Once approved, I'll revert to disabling it by default.

File: src/Interpreters/tests/gtest_filecache.cpp
Comment (danipozo) [reply src/Interpreters/tests/gtest_filecache.cpp]: Done

File: src/Storages/System/StorageSystemFilesystemCache.cpp
Comment (danipozo) [reply src/Storages/System/StorageSystemFilesystemCache.cpp:51]: Done

File: tests/integration/test_filesystem_cache/test.py
Comment (danipozo) [reply tests/integration/test_filesystem_cache/test.py]: Done, now it checks `system.filesystem_cache_settings`

**Review by kssenii — APPROVED** [review]

File: src/Interpreters/TemporaryDataOnDisk.cpp
Comment (danipozo) [reply src/Interpreters/TemporaryDataOnDisk.cpp:68]: Disabled by default

**Comment by kssenii** [comment]
I am going to fix conflicts in "Cloud fork sync", therefore please avoid force pushes, because then I'll have to resolve conflicts manually again.

File: src/Interpreters/Cache/FileCacheSettings.h
Comment (nikitamikhaylov) [reply src/Interpreters/Cache/FileCacheSettings.h:35]: > So we can enable it by default in next version.

@kitaisreal Why do we need it?

**Pull request merged**
